ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -648,68 +672,99 @@ func (r *RayServiceReconciler) updateServeDeployment(ctx context.Context, raySer
-		statusMap[status.Name] = status
-	}
+	r.Log.V(1).Info("getAndCheckServeStatus", "prev statuses", rayServiceServeStatus.Applications, "serve statuses", serveAppStatuses)
 
 	isHealthy := true
 	isReady := true
 	timeNow := metav1.Now()
-	for i := 0; i < len(serveStatuses.DeploymentStatuses); i++ {
-		serveStatuses.DeploymentStatuses[i].LastUpdateTime = &timeNow
-		serveStatuses.DeploymentStatuses[i].HealthLastUpdateTime = &timeNow
-		if serveStatuses.DeploymentStatuses[i].Status != rayv1alpha1.DeploymentStatusEnum.HEALTHY {
-			prevStatus, exist := statusMap[serveStatuses.DeploymentStatuses[i].Name]
-			if exist {
-				if prevStatus.Status != rayv1alpha1.DeploymentStatusEnum.HEALTHY {
-					serveStatuses.DeploymentStatuses[i].HealthLastUpdateTime = prevStatus.HealthLastUpdateTime
-
-					if prevStatus.HealthLastUpdateTime != nil && time.Since(prevStatus.HealthLastUpdateTime.Time).Seconds() > serviceUnhealthySecondThreshold {
+
+	newApplications := make(map[string]rayv1alpha1.AppStatus)
+	for appName, app := range serveAppStatuses {
+		if appName == "" {
+			appName = "default"
+		}
+
+		prevApplicationStatus := rayServiceServeStatus.Applications[appName]
+
+		applicationStatus := rayv1alpha1.AppStatus{
+			Message:              app.Message,
+			Status:               app.Status,
+			LastUpdateTime:       &timeNow,
+			HealthLastUpdateTime: &timeNow,
+			Deployments:          make(map[string]rayv1alpha1.ServeDeploymentStatus),
+		}
+
+		// Check app status
+		if app.Status != rayv1alpha1.ApplicationStatusEnum.RUNNING {
+			// Check previous app status
+			if prevApplicationStatus.Status != rayv1alpha1.ApplicationStatusEnum.RUNNING {
+				if prevApplicationStatus.HealthLastUpdateTime != nil {
+					applicationStatus.HealthLastUpdateTime = prevApplicationStatus.HealthLastUpdateTime
+
+					if time.Since(prevApplicationStatus.HealthLastUpdateTime.Time).Seconds() > serviceUnhealthySecondThreshold {
 						isHealthy = false
 					}
 				}
 			}
 			isReady = false
 		}
-	}
 
-	// Check app status
-	serveStatuses.ApplicationStatus.LastUpdateTime = &timeNow
-	serveStatuses.ApplicationStatus.HealthLastUpdateTime = &timeNow
-	if serveStatuses.ApplicationStatus.Status != rayv1alpha1.ApplicationStatusEnum.RUNNING {
-		// Check previous app status
-		if rayServiceServeStatus.ApplicationStatus.Status != rayv1alpha1.ApplicationStatusEnum.RUNNING {
-			if rayServiceServeStatus.ApplicationStatus.HealthLastUpdateTime != nil {
-				serveStatuses.ApplicationStatus.HealthLastUpdateTime = rayServiceServeStatus.ApplicationStatus.HealthLastUpdateTime
+		// Check deployment statuses
+		for deploymentName, deployment := range app.Deployments {
+			deploymentStatus := rayv1alpha1.ServeDeploymentStatus{
+				Status:               deployment.Status,
+				Message:              deployment.Message,
+				LastUpdateTime:       &timeNow,
+				HealthLastUpdateTime: &timeNow,
+			}
+
+			if deployment.Status != rayv1alpha1.DeploymentStatusEnum.HEALTHY {
```

Comment:
HEALTHY means it has reached the target number of healthy replicas (e.g. if target number of replicas is 3, and there are 3 healthy replicas, Serve marks the deployment as healthy)

Reply:
A replica can be unhealthy if it fails the user-defined health check method consecutively a few times, or if it failed to start. Either way, if a replica becomes unhealthy it will immediately be stopped and restarted.